### DISCUSSION TITLE
PDF to SVG conversion. Colors are changed visually.

### DISCUSSION THREAD

**Discussion by serhii-brovarnyk** [pymupdf/PyMuPDF#3648] [Looking for help] [answered]
Hello!

I have seen some strange results after converting PDF to SVG via `get_svg_image()` method. The original PDF document has lighter colors than the SVG image or PDF documents that I got when converting PyMuPDF's SVG back to PDF via another lib and PyMuPDF as well. So I suppose it is a problem with the PDF to SVG conversion, maybe some unsupported styles or attributes by SVG or something like that.

Is there any way to fix this after conversion? Maybe I can get some information about these attributes from the `get_drawings()` method?

Original PDF file: 
[test4_page_3.pdf](https://github.com/user-attachments/files/16052335/test4_page_3.pdf)

Converted from SVG (another lib):
[pdf_bad_color.pdf](https://github.com/user-attachments/files/16052618/pdf_bad_color.pdf)

Converted from SVG (PyMuPDF):
[pymupdf_bad_color.pdf](https://github.com/user-attachments/files/16052627/pymupdf_bad_color.pdf)

Thank you in advance for any help!

**Comment by JorjMcKie** [comment]
Sorry for the tardy reply!
I cannot confirm your results. Here is what I did:
1. Convert the PDF page to an SVG image via
   ```python
   doc = pymupdf.open("test4_page_3.pdf")
   Path("test.svg").write_text(doc[0].get_svg_image())
   ```
2. Confirmed the correct display of the SVG image in a browser.
3. Converted the the SVG file back to a new PDF page via `mutool draw -o test4_page_3-new.pdf test4_page_3.svg`.
4. Did the same type of conversion SVG-to-PDF via PyMuPDF code
   ```python
   svg=pymupdf.open("test4_page_3.svg")
   pdfdata = svg.convert_to_pdf()
   doc=pymupdf.open("pdf",pdfdata)
   doc.ez_save("test4_page_3-pymu.pdf")
   ```

The results of 3. and 4. look the same (as expected) - except that the PyMuPDF output is much smaller because of the top-compression options used in `.ez_save()`.

However, the result of step 3./4. indeed is not identical to the original PDF page. But the differences are not colors. Instead there are some incorrect reproductions like the following.

Original:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/e41d33c3-cbdc-4646-9b62-6e8464d7d763)
Reproduced:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/21a21697-2731-4bab-ac62-6dd586124a76)

**Reply by serhii-brovarnyk** [reply]
Thank you for the reply!

You should compare colors between the original PDF and PDF based on the SVG image obtained from the original PDF.
There is a picture that represents the difference in appearance between documents:
![image](https://github.com/pymupdf/PyMuPDF/assets/149577482/cc770f75-64c5-4115-9e6a-8e6e3de5443b)
The left part of the picture is a converted document and the right one is the original.
So visually they're slightly different.

It looks like dark-grey colors changed from: `'color': (0.13669031858444214, 0.12195010483264923, 0.1252918243408203)`
To: `'color': (0.13722439110279083, 0.12161440402269363, 0.1254444122314453)`
Colors are obtained via `get_cdrawings()` method.
Maybe I'm wrong because of the absence of verification of whether compared objects are the same or not but seems like this is the issue.
Eventually, I can't find colors from the original PDF file in the converted one so I'm sure that something went wrong during the conversion from PDF to SVG.

**Reply by serhii-brovarnyk** [reply]
And about incorrect reproductions. I believe that this is the issue with handling clip-path elements on the PyMuPDF side. I've seen such a behavior before so generally I use another lib to convert SVG to PDF.

**Comment by JorjMcKie** [comment, marked as answer]
Ah, no: PyMuPDF is out of the loop anyway, because all is done by MuPDF - which PyMuPDF simply invokes.
It appears that we have a conversion problem PDF <-> HTML colors. Your example `color=(0.13669031858444214, 0.12195010483264923, 0.1252918243408203)` for SVG is correctly converted into a HTML color: `stroke="#231f20"`, see here:
```python
b=(0.13669031858444214, 0.12195010483264923, 0.1252918243408203)
[hex(int(round(b[i]*255))) for i in range(3)]
['0x23', '0x1f', '0x20']
```
If you convert that back to PDF, the result is
```python
[int(svgcolor[i], 16)/255 for i in range(3)]
[0.13725490196078433, 0.12156862745098039, 0.12549019607843137]
```
Which is not the exact same value as you extracted it, but still significantly different from the original. So it looks like an inevitable deviation deviation.

**Reply by serhii-brovarnyk** [reply]
Understood. Thank you for your help!
I suppose in case of multiple conversions back and forth color will change even more :(
